src/core/net/ip6.cpp

```diff
@@ -603,27 +622,240 @@ otError Ip6::HandleOptions(Message &aMessage, Header &aHeader, bool &aForward)
+                     error = OT_ERROR_NO_BUFS);
+        VerifyOrExit(aMessage.CopyTo(aMessage.GetOffset() + (offset << 3),
+                                     aMessage.GetOffset() + sizeof(fragmentHeader), payloadFragment,
+                                     *fragment) == (int)payloadFragment,
+                     error = OT_ERROR_NO_BUFS);
+        EnqueueDatagram(*fragment);
+
+        fragmentCnt++;
+        fragment = NULL;
+        otLogInfoIp6("Fragment %d with %d bytes sent ", fragmentCnt, payloadFragment);
+    }
+exit:
+    aMessage.Free();
+
+    if (error == OT_ERROR_NO_BUFS)
+    {
+        otLogWarnIp6("No buffer for Ip6 fragmentation");
+    }
+
+    if (error != OT_ERROR_NONE && fragment != NULL)
+    {
+        fragment->Free();
+    }
+#else
+    OT_UNUSED_VARIABLE(aMessage);
+    OT_UNUSED_VARIABLE(aIpProto);
+#endif // OPENTHREAD_ENABLE_IP6_FRAGMENTATION
+    return error;
+}
+
+otError Ip6::HandleFragment(Message &aMessage, Header &aHeader, MessageInfo &aMessageInfo)
 {
     otError        error = OT_ERROR_NONE;
     FragmentHeader fragmentHeader;
+#if OPENTHREAD_ENABLE_IP6_FRAGMENTATION
+    uint16_t offset          = 0;
+    uint16_t payloadFragment = 0;
+#endif // OPENTHREAD_ENABLE_IP6_FRAGMENTATION
 
     VerifyOrExit(aMessage.Read(aMessage.GetOffset(), sizeof(fragmentHeader), &fragmentHeader) == sizeof(fragmentHeader),
                  error = OT_ERROR_PARSE);
 
+#if OPENTHREAD_ENABLE_IP6_FRAGMENTATION
+    if (fragmentHeader.GetOffset() == 0 && fragmentHeader.IsMoreFlagSet() == false)
+    {
+        error = aMessage.MoveOffset(sizeof(fragmentHeader));
+    }
+    else if (mFragmentBuffer.mFragmentIdentification == 0 ||
+             (fragmentHeader.GetIdentification() == mFragmentBuffer.mFragmentIdentification &&
+              fragmentHeader.GetNextHeader() == mFragmentBuffer.mFragmentNextHeader))
+    {
+        offset          = fragmentHeader.GetOffset() << 3;
+        payloadFragment = aMessage.GetLength() - aMessage.GetOffset() - sizeof(fragmentHeader);
+        mFragmentBuffer.mReassemblyPayloadReceived += payloadFragment;
+
+        // check whether it is the first frame
+        if (mFragmentBuffer.mFragmentIdentification == 0)
+        {
+            // starting the timer for the reassembly timeout
+            mTimer.Start(kIp6ReassemblyTimeoutMilli);
+
+            memcpy(&mFragmentBuffer.mMessageInfo, &aMessageInfo, sizeof(MessageInfo));
+            memcpy(&mFragmentBuffer.mHeader, &aHeader, sizeof(Header));
+
+            mFragmentBuffer.mFragmentIdentification = fragmentHeader.GetIdentification();
+            mFragmentBuffer.mFragmentNextHeader     = fragmentHeader.GetNextHeader();
+
+            VerifyOrExit((mFragmentBuffer.mIp6MsgBuffer = NewMessage(0)) != NULL, error = OT_ERROR_NO_BUFS);
+            SuccessOrExit(error = mFragmentBuffer.mIp6MsgBuffer->SetLength(aMessage.GetOffset()));
+            SuccessOrExit(error = mFragmentBuffer.mIp6MsgBuffer->SetOffset(aMessage.GetOffset()));
```

Comment:
Maybe use `assert()` instead, since a successful call to `SetLength()` should allocate the appropriate message buffers.

Reply:
changed